sidecar/src/connections.rs

```diff
@@ -0,0 +1,171 @@
+use std::{
+    sync::{
+        atomic::{
+            AtomicUsize,
+            Ordering::{Acquire, Relaxed},
+        },
+        Arc,
+    },
+    task::{ready, Poll},
+    time::Duration,
+};
+
+use hyper::server::accept::Accept;
+use pin_project::pin_project;
+use tokio::{
+    io::{AsyncRead, AsyncWrite},
+    net::{UnixListener, UnixStream},
+    time::timeout,
+};
+
+#[pin_project]
+#[derive(Debug)]
+pub struct UnixListenerTracked {
+    listener: UnixListener,
+    connection_tracker: Tracker,
+}
+
+impl UnixListenerTracked {
+    pub fn watch(&self) -> TrackerWatcher {
+        self.connection_tracker.watch()
+    }
+}
+
+impl Accept for UnixListenerTracked {
+    type Conn = UnixStreamTracked;
+
+    type Error = std::io::Error;
+
+    fn poll_accept(
+        self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Option<Result<Self::Conn, Self::Error>>> {
+        let stream = ready!(self.listener.poll_accept(cx))?.0;
+        Poll::Ready(Some(Ok(UnixStreamTracked {
+            inner: stream,
+            tracker: self.connection_tracker.clone(),
+        })))
+    }
+}
+
+impl From<UnixListener> for UnixListenerTracked {
+    fn from(listener: UnixListener) -> Self {
+        Self {
+            listener,
+            connection_tracker: Tracker::default(),
+        }
+    }
+}
+
+#[pin_project]
+pub struct UnixStreamTracked {
+    #[pin]
+    inner: UnixStream,
+    tracker: Tracker,
+}
+
+impl AsyncWrite for UnixStreamTracked {
+    fn poll_write(
+        self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+        buf: &[u8],
+    ) -> Poll<Result<usize, std::io::Error>> {
+        self.project().inner.poll_write(cx, buf)
+    }
+
+    fn poll_flush(
+        self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> Poll<Result<(), std::io::Error>> {
+        self.project().inner.poll_flush(cx)
+    }
+
+    fn poll_shutdown(
+        self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> Poll<Result<(), std::io::Error>> {
+        self.project().inner.poll_shutdown(cx)
+    }
+}
+
+impl AsyncRead for UnixStreamTracked {
+    fn poll_read(
+        self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+        buf: &mut tokio::io::ReadBuf<'_>,
+    ) -> Poll<std::io::Result<()>> {
+        self.project().inner.poll_read(cx, buf)
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct Tracker {
+    count: Arc<AtomicUsize>,
+    notifier: Arc<tokio::sync::Notify>,
+}
```

Comment:
One arc is better than two maybe?

```suggestion
struct TrackerInner {
    count: AtomicUsize,
    notifier: tokio::sync::Notify,
}

pub struct Tracker {
  inner: Arc<TrackerWatcherInner>
}
```

Reply:
I need to be able to clone either the counter or notifier separately